Howudong Bridge-Game 미션 제출합니다.

howudong opened the pull request:
오래 걸렸네요.. 데드라인 넘겼네요..

howudong committed:
docs : 기능 및 입출력 명세 작성

howudong committed:
refactor : domains 패키지 추가 및 배치

howudong committed:
docs : 기능 명세 사항 변경

howudong committed:
feat : 다리 생성 기능 추가 및 BridgeMap에서 이 값을 저장하도록 함

- BridgeMaker를 통해 만들어진 맵을 BridgeMap 생성자로 보냄

howudong committed:
refactor : public 상수 캡슐화를 위해 private으로 변경

howudong committed:
test : BridgeMaker 테스트 코드 작성

howudong committed:
feat : retry를 할때마다 게임 횟수를 기록하는 기능 추가

howudong committed:
feat : 입력 메시지 출력 추가

howudong committed:
feat : 다리 길이를 입력받는 기능 추가

howudong committed:
feat : 입력이 올바른지 체크하는 InputChecker 클래스 생성

howudong committed:
feat : 생성된 다리 정보와 움직인 경로, 해당 방향이 맞는지 확인하는 함수를 가진 BridgeMap 생성

howudong committed:
refactor : 예외처리를 makeBridge를 호출한 클래스에서 처리하도록 예외 처리 제거 및 생성자 변경

howudong committed:
feat : 게임 정보를 가지고 있는 dto 생성

howudong committed:
feat : 게임 시작 후 사이즈 입력까지 받게 구현

howudong committed:
refactor : 우테코 요구사항대로 다시 원상복구

howudong committed:
docs : 필요없는 도메인 삭제 및 세부사항 갱신

howudong committed:
refactor : inputView에서는 입력이 정수인지만 확인하도록 범위 확인하는 함수는 삭제

howudong committed:
test : 함수 구조 변경에 따른 test 코드 새로 만듦

howudong committed:
docs : 중복되는 요구사항 제거

howudong committed:
refactor : inputView로 커맨드 관련 상수 옮김

howudong committed:
feat : 커맨드가 정상적인지 확인하는 validateCommand 기능 구현

howudong committed:
style : startGame에서 play로 함수 이름 변경

howudong committed:
test : valdiateCommand의 테스트 케이스 추가

howudong committed:
feat : 공개해도 되는 정보만 모아둔 GameInformationDto 생성

howudong committed:
refactor : final 위치 바꿈

howudong committed:
feat : 컨트롤러와 view, domain에 dto로 변환하여 사용할 수 있ㄱ ㅔ하는 BridgeService 생성

howudong committed:
refactor : isGameSuccess 변수 default로 바꿔 BridgeMap에서 조작할 수 있도록 허용

howudong committed:
feat : 모든 종류의 출력문 함수 만듦

howudong committed:
refactor: 예외 처리 및 예외발생시 해당 함수 재시작하는 것을 inputview에서 하도록 처리

howudong committed:
feat : dto에 필요한 getter 생성

howudong committed:
style : moveHistory 함수를 제거하고, 리스트로 받게함

howudong committed:
feat : 다리를 다 건널때까지 다리 건너기를 반복하는 play() 함수 구현

howudong committed:
feat : 서비스와 컨트롤러 연동

howudong committed:
fix : 다리 크기를 잘못입력했을때 제대로 실행되지 않았던 오류 수정

howudong committed:
refactor : 캡슐화를 위해 reset은 bridgeMap에서 호출하도록 함

howudong committed:
refactor : 의미없는 개행 삭제 및 주석 삭제

nuyh99 commented on src/main/java/bridge/Application.java:
```diff
@@ -3,6 +3,6 @@
 public class Application {
 
     public static void main(String[] args) {
-        // TODO: 프로그램 구현
+        new Controller();
```
![](https://github.com/nuyh99/java-bridge/assets/93072571/ddd41ef1-371e-4de3-a203-f86287e7e4aa)
1. 확인용 콘솔 출력을 지우지 않은 것 같네요
2. 생성자를 호출하기만 하는데 프로그램이 실행되는 것은 어색한 것 같습니다. 생성자가 무엇인가요?

howudong replied:
1. 테스트 여부를 하다보니 미처 지우지 못한 것이 있었습니다. 수정했습니다.

2. 
객체지향 관점에서 생성자의 역할에 대해서는 한번도 생각해보지 않은 것 같습니다.
그저 불필요한 코드를 줄이는 정도로 이해하고 있었는데,
말씀대로 생성자로 프로그램 자체가 실행되는 것은 문제가 있는 것 같습니다.

여럿 찾아보니, 객체지향에서 생성자는 의미적인 측면에서 더 중요한 역할을 하는 것 같습니다.
생성자를 통해 객체의 인스턴스가 생성됨과 동시에 값을 가짐으로써 프로그램 내에서 유의미한 엔티티가 될 수 있는 것 같습니다.

그래서 생성자는 어떤 객체가 유의미한 객체가 될 수 있도록 하는 역할을 한다고 생각합니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+    private final OutputView outputView;
+    private final BridgeService bridgeService;
+
+    public Controller() {
+        inputView = new InputView();
+        outputView = new OutputView();
```
컨트롤러가 내부 필드로 View들을 가지고 있는 상태입니다.
현재 `Controller`의 생성자 내부에서 해당 필드들을 초기화하고 있는데 다음과 같은 상황에서 문제될 수 있을 것 같습니다.
- 테스트용 View를 사용해서 Controller를 테스트하고 싶다.
```java
new Controller(new TestInputView(), new TestOutputView());
```
위와 같은 형태로는 가능할 것 같은데 지금의 구조와 어떤 차이가 있을까요?

howudong replied:
새롭게 만든 InputView, OutputView class에 대한 테스트 가능성에 대해서는 생각해보지 않았네요.
생성자를 잘 이용하면, 캡슐화된 부품을 잘 이용할 수 있을 것 같습니다.

제가 만든 방식은 OutputView와 InputView 클래스 자체와 너무 높은 의존도를 가지고 있는 것 같습니다.
그래서 다른 클래스 파일을 사용하고 싶을 때 일일이 수정해야하는 귀찮음이 생길 것 같습니다.

새롭게 수정한 방식은 부품을 바꿔끼우듯, 클래스와의 낮은 결합도를 보이는 것 같습니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+
+        BridgeMap bridgeMap = makeBridge();
+        bridgeService = new BridgeService(bridgeMap, new BridgeGame(bridgeMap));
+        play();
```
생성자의 역할은 무엇일까요?
사용자의 입장에서는 인스턴스만 생성하고 싶은데 내부 로직이 수행되는 것 같아요.

howudong replied:
- 생성자의 기본적인 용도
    - 객체 인스턴스를 생성할 때, 인스턴스 안에 변수를 임의로 할당해서 생성하기 위해 사용
- 객체지향 프로그래밍에서 생성자의 의미는 좀 더 중요한 역할을 담당
→ **생성자는 어떤 객체가 생성과 동시에 의미를 가지는 객체라는 것을 알리는 역할**

제가 수집한 정보로는 이정도로 이해했습니다.

제 생각에는, 생성자는 객체라는 것을 만들어주기만 할 뿐 어떠한 동작을 하면 안된다고 생각합니다.

그래서 위 코드와 더불어 제가 짠 대부분의 코드가 생성자 내부에 다른 함수가 많이 들어가 있어 이를 수정해야할 것 같습니다.

howudong replied:
모든 생성자에서 필요한 전역변수 설정 이외에는 다 없앴습니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+        finishGame(bridgeService.checkBridge());
+    }
+
+    private BridgeMap makeBridge() {
+        BridgeMap bridgeMap;
+        int readSize = inputView.readBridgeSize();
+
+        try {
+            bridgeMap = new BridgeMap(new BridgeMaker(new BridgeRandomNumberGenerator()), readSize);
+        } catch (IllegalArgumentException e) {
+            System.out.println(e.getMessage());
+            bridgeMap = makeBridge();
+        }
+        return bridgeMap;
+    }
```
```suggestion
private BridgeMap makeBridge() {
        int readSize = inputView.readBridgeSize();

        try {
            return new BridgeMap(new BridgeMaker(new BridgeRandomNumberGenerator()), readSize);
        } catch (IllegalArgumentException e) {
            System.out.println(e.getMessage());
            return makeBridge();
        }
    }
```
1. 재귀적으로 재입력 받기 위함이라면 위와 같이 쓸 수 있을 것 같네요.
2. System.out.println은 출력문입니다. OutputView의 역할이 아닐까요? 컨트롤러가 직접 UI 동작을 하면 어떤 문제가 있을까요?

howudong replied:
2.
OutputView가 있음에도 controller가 view의 역할을 수행하면 
controller가 2가지 역할을 수행하는 것이므로 mvc 패턴의 의미가 사라집니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void questionRestart(GameInformationDto gameInformationDto) {
+        if (gameInformationDto.isSuccess()) return;
```
중괄호 생략은 컨벤션이 아닙니다!

howudong replied:
습관적으로 그러네요 수정하도록 하겠습니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+
+        String readCommand = inputView.readGameCommand();
+
+        if (readCommand.equals(InputView.RESTART_COMMAND)) {
```
readCommand는 null이 될 수도 있습니다.
되도록 null의 가능성이 없는 객체의 메서드를 호출하는 것이 좋습니다.
```java
InputView.RESTART_COMMAND.equals(readCommand);
```

---

**아래는 지금 당장하라는 건 아니고 생각해볼 부분입니다.**
추가적으로 컨트롤러 내부에 로직이 있습니다.
이것도 검증해야 하지 않을까요? 로직들을 최대한 하위 레이어(Domain 등)로 위치시키려는 이유는 무엇일까요?

howudong replied:
```
 static void validateNumeric(String readSize) {
        if (readSize == null)
            throw new IllegalArgumentException("[ERROR] 숫자만 입력해주세요");

        try {
            Integer.parseInt(readSize);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("[ERROR] 숫자만 입력해주세요");
        }
    }

    static void validateCommand(List<String> commands, String readDirection) {
        if (readDirection == null
                || commands.stream()
                .anyMatch(e -> e.equals(readDirection))) {
            throw new IllegalArgumentException("[ERROR] 올바르지 않은 입력입니다.\n" + commands + " 중 하나를 입력해주세요.");
        }
    }
```

읽은 값에 대한 null 값인지에 대한 조건문으로 처리를 해줬습니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+        if (readCommand.equals(InputView.RESTART_COMMAND)) {
+            bridgeService.restartGame(readCommand);
+            play();
+            return;
```
필요 없는 구문인 것 같습니다.

howudong replied:
후에 inputview에서 재귀적으로 호출되겠금 햇는데 전에 기능을 삭제 안했네요..

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+
+    private void finishGame(GameInformationDto gameInformationDto) {
+        boolean isSuccess = gameInformationDto.isSuccess();
+        List<String> moveHistory = gameInformationDto.getMoveHistory();
```
moveHistories가 좀 더 맞지 않을까요?!

howudong replied:
움직임 기록이라는 뜻으로 써놨었는데, List는 복수형으로 쓰는게 맞는것 같습니다.
근데 histories라 쓰면 움직임 기록들의 집합체로 느껴질 수 있을 것 같아

함수명과 변수명을 전체적으로 moveLogs로 수정했습니다.

nuyh99 commented on src/main/java/bridge/Controller.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void play() {
+        GameInformationDto bridgeMapDto = bridgeService.checkBridge();
```
1. DTO를 쓰셨는데요, 쓰신 이유가 궁금합니다.
2. DTO 내부에 도메인 객체를 직접 참조해도 될까요?

howudong replied:
1. DTO를 이용하면 컨트롤러에게 공유하고 싶은 데이터만을 선별해서 모아둠으로써, 
프로그래머로 하여금, 도메인에서 어떤 데이터가 사용 가능한지를 명시할 수 있습니다.

2, 1번의 이유로 DTO에 들어간 데이터는 애초에 컨트롤러에게 공개하기 위해 모은 데이터의 클래스입니다.
하지만, 직접 참조는 값을 수정할 가능성이 있기 때문에 private으로 수정을 막아둔 채
get을 사용하여 가져오게만 해야합니다.

nuyh99 commented on src/main/java/bridge/services/BridgeService.java:
```diff
@@ -0,0 +1,32 @@
+
+public final class BridgeService {
+    private final BridgeMap bridgeMap;
+    private final BridgeGame bridgeGame;
```
1. 서비스 레이어가 상태를 가졌을 때의 문제는 무엇일까요?
2. 현재 컨트롤러가 서비스를 필드로 가지고 있기 때문에 컨트롤러 또한 상태를 가진 컨트롤러가 됐습니다. 문제가 있을까요?

nuyh99 commented on src/main/java/bridge/services/BridgeService.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public void restartGame(String command) {
+        bridgeGame.retry(command);
```
`command`는 컨트롤러에서만 쓰이는 것 같습니다.

howudong replied:
생각해보니 필요없는 매개변수더라구요 삭제했습니다!

nuyh99 commented on src/main/java/bridge/views/InputChecker.java:
```diff
@@ -0,0 +1,27 @@
+
+public final class InputChecker {
+
+    private static final int SIZE_LOWER_BOUND = 3;
+    private static final int SIZE_UPPER_BOUND = 20;
+
+
```
쓰이지 않는 상수와 두 줄 개행이 존재하는 것 같네요!

howudong replied:
꼼꼼히 체크하겠습니당..

nuyh99 commented on src/main/java/bridge/views/InputChecker.java:
```diff
@@ -0,0 +1,27 @@
+    private static final int SIZE_UPPER_BOUND = 20;
+
+
+    static void validateNumeric(InputView inputView, String readSize) {
```
InputView가 쓰이나요??
인텔리제이를 사용한다면 쓰이지 않는 변수들은 회색으로 표시됩니다.

howudong replied:
프로그램 완성 후 전체적으로 회색 변수를 보는 습관을 들여야겠네요ㅣ..

nuyh99 commented on src/test/java/bridge/views/InputCheckerTest.java:
```diff
@@ -0,0 +1,53 @@
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class InputCheckerTest {
```
UI로 생각하고 뺄 수 있는 부분인데 테스트 굿👍

howudong replied:
ㅎㅎ

nuyh99 commented on src/main/java/bridge/views/InputView.java:
```diff
@@ -0,0 +1,58 @@
+            readCommand = readGameCommand();
+        }
+        return readCommand;
+    }
```
```suggestion
    public String readGameCommand() {
        outputView.printRestartQuestion();
        String readCommand = Console.readLine();

        try {
            InputChecker.validateCommand(List.of(QUIT_COMMAND, RESTART_COMMAND), readCommand);
            return readCommand;
        } catch (IllegalArgumentException e) {
            System.out.println(e.getMessage());
            return readGameCommand();
        }
    }
```
1. 위와 같이 리턴을 하도록 바꿀 수 있을 것 같습니다.
2. System.out.println이 컨트롤러에도 존재하고 InputView에도 존재하네요. `System.out#println`은 출력의 구현 중 한 방법이라고 생각되는데 OutputView에만 존재하도록 하면 더 좋지 않을까요? 어떤 장단점이 있을까요?

howudong replied:
1. 
함수의 리턴 타입은 무조건 밖에 마지막 블록에 있어야 하는건줄 알았는데 
try 블록과 catch 블록에 하나씩 있으면 가능한지 처음 알았네요.

덕분에 많이 알아갑니다.

2.
처음에는 domain에서 예외처리를 하다보니 
당연히 view와 연관되면 안되기 때문에 출력을 따로 했는데,

생각해보니 충분히 가능하네요.

해당 방식의 장점은 모든 출력 관련된 문구는 OutputView class에 있으므로
타인이 코드의 위치를 찾기 편해지고, 수정도 편해집니다.

단점은 음 굳이.. OutputView의 의존성이 높아지는 것 정도..?

nuyh99 commented on src/main/java/bridge/views/OutputView.java:
```diff
@@ -0,0 +1,88 @@
+
+    private String drawSection(String keyWord, List<String> moveHistory, boolean isSuccess) {
+        List<String> printWords = moveHistory.stream()
+                .map((e) -> getPrintWord(keyWord, e))
```
e는 무엇을 뜻하는 건가요??

howudong replied:
히스토리 로그를 뜻하는 거였는데.. 정신나가서 코딩하다보니 맛간듯

nuyh99 commented on src/main/java/bridge/views/OutputView.java:
```diff
@@ -0,0 +1,88 @@
+                .map((e) -> getPrintWord(keyWord, e))
+                .collect(Collectors.toList());
+
+        convertLastWord(printWords, moveHistory.size() - 1, isSuccess);
+        return String.join(" | ", printWords);
```
딱 봤을 때는 왜 -1을 하는 지 알 수 없을 것 같습니다. 어떤 방법이 있을까요?
" | " 이건 각 스텝 별 사이 구간 출력문인 것 같은데 상수로 치환하면 더 알기 쉽고 관리하기도 편할 것 같습니다.

howudong replied:
```
 private String drawSection(String keyWord, List<String> moveHistory, boolean isSuccess) {
        ...
        convertLastWord(printWords, isSuccess); // 매개변수 없앰
        return String.join(BRIDGE_DELIMITER, printWords);
    }

    private void convertLastWord(List<String> printWords, boolean isSuccess) {
        int lastIndex = printWords.size() - 1; // 알아보기 쉽게 추가
        if (!isSuccess && printWords.get(lastIndex).equals(SUCCESS_SIGN)) {
            printWords.set(lastIndex, FAIL_SIGN); // 그냥 set으로 하면 되더라구요
        }
    }
```

nuyh99 commented on src/main/java/bridge/views/OutputView.java:
```diff
@@ -0,0 +1,88 @@
+
+    private void convertLastWord(List<String> printWords, int length, boolean isSuccess) {
+        System.out.println(isSuccess);
+        if (!isSuccess && printWords.get(length).equals("O")) {
+            printWords.remove(length);
+            printWords.add("X");
```
O, X도 마찬가지로 치환하면 더 좋을 것 같습니다.
비즈니스적으로 바뀔 수 있는 출력 형식이라면 상수로 관리하는 것이 메서드를 볼 때도, 값을 바꾸기도 쉬울 것 같네요.

howudong replied:
O,X는 여기서밖에 쓰이지 않아서 따로 상수처리를 하지 않았는데,
비즈니스적으로 변경될 가능성을 고려하지 않았네요..!

상수로 따로 뺐습니다!

nuyh99 commented on src/main/java/bridge/dto/GameInformationDto.java:
```diff
@@ -0,0 +1,31 @@
+    public boolean isSuccess() {
+        return bridgeGame.isGameSuccess();
+    }
+
+
```
# 👀

howudong replied:
꼼꼼히 체크하겠습니다..

nuyh99 commented on src/main/java/bridge/domains/BridgeRandomNumberGenerator.java:
```diff
@@ -1,8 +1,8 @@
 import camp.nextstep.edu.missionutils.Randoms;
 
-public class BridgeRandomNumberGenerator implements BridgeNumberGenerator {
+public final class BridgeRandomNumberGenerator implements BridgeNumberGenerator {
```
상속 금지 final 👍
근데 그렇다면 상속할 이유가 없는 모든 클래스에 붙여야 하지 않을까요?
DTO, Application 등에는 빠져있는 것 같습니다.

+ 이 클래스도 테스트해볼 수 있을 것 같습니다.

howudong replied:
해당 테스트 코드를 작성하는 과정에서는 필요가 없어서 제외했지만
랜덤 테스트 코드 작성에 대해 알아보면서 

함수형 인터페이스, 메서드 참조, 
여기서 BridgeNumberGenerator라는 인터페이스가 있는 이유  등에 대해 알게됐습니다. 

많은 공부가 됏네요

nuyh99 commented on src/test/java/bridge/domains/BridgeMakerTest.java:
```diff
@@ -0,0 +1,26 @@
+
+class BridgeMakerTest {
+
+    @DisplayName("다리 크기가 3미만이면 예외를 발생시킨다.")
+    @Test
+    void validateBridgeToLowerBound() {
+        Assertions.assertThatThrownBy(() ->
+                        new BridgeMaker(new BridgeRandomNumberGenerator())
+                                .makeBridge(2))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```
```suggestion
    @DisplayName("다리 크기가 3미만이면 예외를 발생시킨다.")
    @Test
    void validateBridgeToLowerBound() {
        final BridgeMaker bridgeMaker = new BridgeMaker(new BridgeRandomNumberGenerator());

        assertThatThrownBy(() -> bridgeMaker.makeBridge(2))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("[ERROR] 다리 길이는 3 이상 20 이하만 가능합니다.");
    }
```
현재 `IllegalArgumentException`이 발생하는 부분이 많은 것 같습니다.
`hasMessage()`로 더 범위를 좁혀주는 것이 좋습니다.
검증하려는 예외만 잡도록 작성하는 것이 좋습니다.

Assertions를 static import하면 더 깔끔해집니다.
Assertion 부분은 최대한 간결하게 알아볼 수 있어야 합니다.

howudong replied:
hasMessage로 좀 더 구체화해야한다는 걸 인지하지 못했습니다.

또한, 방향 입력과 게임 재시작 입력 예외에 대한 테스트도 별도로 처리하도록 했습니다.

nuyh99 commented on src/main/java/bridge/domains/BridgeGame.java:
```diff
@@ -0,0 +1,32 @@
+package bridge.domains;
```
패키지명은 단수로 하는 것이 좋습니다.

howudong replied:
수정했습니다!

nuyh99 commented on src/main/java/bridge/domains/BridgeGame.java:
```diff
@@ -0,0 +1,32 @@
+package bridge.domains;
+
+public final class BridgeGame {
+    private final BridgeMap bridgeMap;
+    private static int gameTryCount;
+    static boolean gameSuccess;
```
default 접근제어자를 쓴 것은 좋습니다.
다만 `gameSuccess`를 사용하는 `BridgeMap#isFinishBridge`에서 파라미터로 BridgeGame 인스턴스를 받아서 체크해볼 수도 있을 것 같습니다.
이번에 적용하라는 것은 아니지만 현재의 구조와 어떤 차이가 있을지, 파라미터로 받으려면 어떻게 리팩토링해야 할 지 생각해보면 좋을 것 같네요

howudong replied:
제가 했던 방식을 사용하면 같은 패키지 안에서라고는 해도,
 gameSuccess 정보가 수정가능해져서 캡슐화가 깨져버리네요.

그래서 매개변수 방식을 생각해봤는데, 이는 클래스간의 결합도를 높힐 것 같습니다.

왜냐하면 BridgeMap에서는 BridgeGame을 참조하고 있지 않습니다.

특히 BridgeGame이 이미 BridgeMap을 참조하고 있어서, 
만약 BridgeGame을 참조해버리면 쌍방 참조가 일어나 결합도가 엄청나게 높아질 것 같습니다.

그래서 이미 BridgeMap을 참조하고 있는 쪽에서 결과를 확인하는 함수를 호출하는 방식으로 바꾸었습니다.

BridgeGame.class

```
public boolean isGameSuccess() {
        return bridgeMap.isFinishBridge();
    }
```

BridgeMap.class

```
public boolean isFinishBridge() {
        return (moveLogs.size() == map.size())
                && (moveLogs.toString().equals(map.toString()));
    }
```

nuyh99 commented on src/main/java/bridge/domains/BridgeMap.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public final class BridgeMap {
```
도메인에 위치한 클래스들은 모두 쉽게 테스트 작성이 가능하지 않을까요?
도메인 클래스들의 모든 메서드, 모든 분기가 테스트돼야 할 것 같습니다.
테스트 코드의 역할은 무엇인가요?

howudong replied:
테스트를 하는 과정에서 BridgeGame class 와 BridgeMap 클래스가 분리되어있는 것이 의미없다고 느껴졌습니다.
왜냐하면 Game 쪽에서 BridgeMap의 객체를 너무 많이 사용하고, Game 쪽에 있어도 될법한게 Map에 있어서 입니다.

그래서 두 클래스를 하나의 BridgeGame으로 합쳤습니다.

howudong replied:
테스트 코드의 역할은 작성한 코드가 예외사항 없이 정상적인 의도대로 돌아가는지를 테스트해보기 위해 있는 것 같습니다.
테스트 코드를 씀으로써, 코드 에러 수정의 시간을 아낄 수 있습니다.

그리고 테스트하기 쉬운 코드를 짤려다보니 자연스럽게 객체지향적으로 함수를 짜게 되는 것 같습니다.

nuyh99 commented on src/test/java/bridge/domains/BridgeMakerTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class BridgeMakerTest {
```
에러 케이스에 대해서는 모두 검증을 잘 했습니다.
정상 케이스는 테스트되지 않은 것 같네요!

howudong replied:
정상 케이스를 검증하는 과정에서 랜덤 넘버에 테스트 코드에 대한 많은 공부가 됐습니다.

nuyh99 commented on src/test/java/bridge/views/InputCheckerTest.java:
```diff
@@ -0,0 +1,53 @@
+                -> InputChecker.validateCommand(List.of("D", "U"), "A"))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+
```
# 👀

howudong replied:
테스트 로직이라고 생각하고 컨벤션을 지키지 않았던 것 같습니다..

nuyh99 reviewed (changes requested):
전체적으로 컨벤션을 잘 지킨 것 같습니다.
하지만 아직 못 지킨 부분도 존재하는 건 사실이고, 테스트의 수가 부족한 것 같습니다.
다른 것들도 중요하지만 현재 단계에서는 컨벤션 100% 지키기와 작성 가능한 모든 테스트를 작성하는 것부터 공부하는 것이 좋겠네요! JUnit과 AssertJ를 사용해서 모든 경우를 테스트할 수 있으니 한 번 고민해보시죠!

howudong committed:
fix : 출력으로 디버깅 로그가 나오는 것을 수정

howudong committed:
refactor : 의미 없는 개행 삭제

howudong committed:
refactor : 수정 가능성 있는 출력 문자 상수 처리

howudong committed:
refactor : 좀 더 명시적으로 convertLastWord 함수 변경

howudong committed:
test : 테스트 케이스 구체화
- 에러 메시지로 어떤 커맨드를 입력해야하는지 나오게함
- 방향 입력과 게임 재시작 입력의 테스트를 따로 분리

howudong committed:
style : 잘못됐던 혹은, 애매했던 이름 변경

- domains -> domain
- getMoveHistory() -> getMoveLogs()
- moveHistory -> moveLogs
- e -> currentLog

howudong committed:
refactor: 필요없는 구문 삭제

howudong committed:
refactor: 클래스의 할 일을 명확히 하기 위해서 OutputView에 printErrorMessage 함수 생성 및 InputView 리팩토링

howudong committed:
refactor : outputview 역할로 하는 것이 맞는 것 같아 OutputView에서 호출하도록 함

howudong committed:
refactor : outputview 역할로 하는 것이 맞는 것 같아 OutputView에서 호출하도록 함

howudong committed:
refactor : outputview 역할로 하는 것이 맞는 것 같아 OutputView에서 호출하도록 함

howudong committed:
refactor : 불필요한 매개변수 삭제

howudong committed:
refactor : final class로 변경

howudong committed:
test : generate()에 대한 테스트 코드 작성

howudong committed:
refactor : 불필요한 주석 제거

howudong committed:
fix : 입력에서 공백을 입력했을 때, 정상 진행이 되던 문제를 수정

howudong committed:
fix : 잘못된 출력 메시지 수정

howudong committed:
fix : 게임이 재시작되지 않았던 기능 누락 사항 구현

howudong committed:
fix : R,Q가 포함되면 실패했던 문제 수정

howudong committed:
style : Test 함수 이름 변경

howudong committed:
refactor : 도메인 클래스간 결합도를 낮추기 위해 메서드 수정

howudong committed:
refactor : 높은 의존도로, 가독성을 해쳐서 BridgeMap.jav 클래스 삭제 및 BridgeGame으로 통합

- 의미상 BridgeGame에 있는게 맞는거 같기도 합니다
- 게임에 맵 정보와 이동 기록이 저장되는게 맞는 것 같다.

howudong committed:
test : BridgeGame에 대한 필요한 테스트 코드 작성

- move 함수와 isGameSuccess에 대해 작성
- 성공과 실패의 경우를 모두 테스트함

howudong committed:
fix : 출력 명세사항 지켜지지 않아서 수정

howudong committed:
feat : 생성된 다리만을 관리 및 교환을 위한 MapDto 생성

howudong committed:
feat: dto를 통해 BridgeGame 객체를 만들 수 있도록 생성자 오버로딩

howudong committed:
fix: MoveLogs 리스트에 추가가 불가능하여 unmodifiableList를 삭제

howudong committed:
refactor : 자바 컨벤션에 맞춰 코드 리팩토링

howudong committed:
refactor: Controller가 서비스와 dto 기반으로 동작하도록 코드 리팩토링

howudong committed:
refactor: dto와 model 간의 변환을 포함한 dto를 사용하도록 코드 리팩토링

- step -> stepNext로 이름 변경하고 GameDto를 반환
- BridgeGame에 있던 isMatch 이동
- makeBridgeMap -> 다리를 만드는 함수 여기로 이동

howudong committed:
test : 변경된 BridgeGame에 맞는 테스트
